flang/include/flang/Lower/Support/BoxValue.h

```diff
@@ -209,6 +211,86 @@ class BoxValue : public AbstractBox, public AbstractArrayBox {
 /// Used for triple notation (array slices)
 using RangeBoxValue = std::tuple<mlir::Value, mlir::Value, mlir::Value>;
 
+/// BoxAddressValue is used for entities that are represented by the address of
+/// a box. This is intended to be used for entities whose base address, shape
+/// and type are not constant in the entity lifetime (e.g Allocatables and
+/// Pointers).
+class BoxAddressValue : public AbstractBox {
+public:
+  /// Create BoxAddressValue given the address \p addr of the box and the non
+  /// deferred length parameters \p lenParameters. The non deferred length
+  /// parameters must always be provided, even if they are constant and already
+  /// reflected in the address type.
+  BoxAddressValue(mlir::Value addr, llvm::ArrayRef<mlir::Value> lenParameters)
+      : AbstractBox(addr), lenParams{lenParameters.begin(),
+                                     lenParameters.end()} {
+    // Currently only accepts fir.(ref/ptr/heap)<fir.box<type>> mlir::Value for
+    // the address. This may change if we accept
+    // fir.(ref/ptr/heap)<fir.heap<type>> for scalar without length parameters.
+    assert(verify() &&
+           "BoxAddressValue requires mem ref to fir.box<fir.[heap|ptr]<type>>");
+  }
+
+  /// Get the fir.box<type> part of the address type.
+  fir::BoxType getBoxTy() const {
+    auto type = getAddr().getType();
+    return fir::dyn_cast_ptrEleTy(type).cast<fir::BoxType>();
+  }
+  /// Return the part of the address type after memory and box types. That is
+  /// the element type, maybe wrapped in a fir.array type.
+  mlir::Type getBaseTy() const {
+    return fir::dyn_cast_ptrEleTy(getBoxTy().getEleTy());
+  }
+  /// Get the scalar type related to the described entity
+  mlir::Type getEleTy() const {
+    auto type = getBaseTy();
+    if (auto seqTy = type.dyn_cast<fir::SequenceType>())
+      return seqTy.getEleTy();
+    return type;
+  }
+
+  /// Is this a Fortran pointer ?
+  bool isPointer() const {
+    return getBoxTy().getEleTy().isa<fir::PointerType>();
+  }
+  /// Is this an allocatable ?
+  bool isAllocatable() const {
+    return getBoxTy().getEleTy().isa<fir::HeapType>();
+  }
+  /// Is the entity an array or an assumed rank ?
+  bool hasRank() const { return getBaseTy().isa<fir::SequenceType>(); }
+  /// Is this an assumed rank ?
+  bool hasAssumedRank() const {
+    auto seqTy = getBaseTy().dyn_cast<fir::SequenceType>();
+    return seqTy && seqTy.hasUnknownShape();
+  }
+  /// Returns the rank of the entity. Beware that zero will be returned for
+  /// both scalars and assumed rank.
+  unsigned rank() const {
+    auto seqTy = getBaseTy().dyn_cast<fir::SequenceType>();
+    if (seqTy)
+      return seqTy.getDimension();
+    return 0;
+  }
+  /// Is this a character entity ?
+  bool isCharacter() const { return getEleTy().isa<fir::CharacterType>(); };
+  /// Is this a derived type entity ?
+  bool isDerived() const { return getEleTy().isa<fir::RecordType>(); };
```

Comment:
These methods are the centralizing the assumption regarding what is inside a `MutableBoxValue`. The assumption is that for scalars, we may want to get rid of the box, in which case I want to minimize the places that needs to know if the address contains a box or not to get to the answer is this a charatcter and such.
Since they operate on possible combinations of fir types, rather than on a single fir types, and are based on assumptions of what the address type is (verified at `MutableBoxValue` construction), I feel they belong better here.